src/datahike/writing.cljc

```diff
@@ -147,27 +147,16 @@
                   (<?- branch-op)
                   db)))))
 
-(defn add-commit-meta! [connection commit-db]
-  (let [{{:keys [datahike/commit-id datahike/parents]} :meta} commit-db]
-    (swap! connection #(-> %
-                           (assoc-in [:meta :datahike/parents] parents)
-                           (assoc-in [:meta :datahike/commit-id] commit-id)))))
-
-(defn update-connection! [connection tx-data tx-meta update-fn]
-  (let [ret-atom (atom nil)]
-    (swap! connection
-           (fn [old]
-             (let [{:keys [writer]} old
-                   {:keys [db-after]
-                    {:keys [db/txInstant]}
-                    :tx-meta
-                    :as   tx-report} (update-fn old tx-data tx-meta)
-                   new-meta               (assoc (:meta db-after) :datahike/updated-at txInstant)
-                   db                     (assoc db-after :meta new-meta :writer writer)
-                   tx-report              (assoc tx-report :db-after db)]
-               (reset! ret-atom tx-report)
-               db)))
-    @ret-atom))
+(defn update-db [old tx-data tx-meta update-fn]
```

Comment:
I am thinking whether it would make sense to structure this code slightly differently. This function, `update-db`, is for example called like this:

```clj
(update-db old entities nil #(core/load-entities-with %1 %2 %3)))
```

and `update-db` simply calls the `update-fn` on the provided args and does some wrangling of the datastructures. Would it make sense to move that step *outside* of this function, and rename this function to `complete-db-update`? If so, the above line would become

```clj
(complete-db-update (core/load-entities-with old entities nil))
```

and `complete-db-update` could simply be something like

```clj
(defn complete-db-update [new-db]
  (-> new-db
      (assoc-in [:db-after :meta :datahike/updated-at]
                (-> new-db :tx-meta :db/txInstant))
      (assoc-in [:db-after :writer]
                (:writer new-db))))
```

Likewise, we would also have 

* `(complete-db-update (core/with old tx-data tx-meta))`
* `(complete-db-update (core/with old entities nil))`
* `(complete-db-update (with @conn tx-data tx-meta))`

and so on.

Reply:
Thanks for forcing me to clean this up a bit, it is definitely better without higher order functions like this.

Reply:
Thanks :-)